Salt-SSH JSON render failed - dumping password as string in output?

Hi,

Bit of a strange one here and I've not been able to get to the bottom of it - or figured out how to reliably replicate it.

Salt-SSH occasionally gets itself into a state where it will randomly output the password used for SSHing to the server into the JSON output it's trying to render, causing a render failure. When it's doing this it will always output the password at the end of the JSON output, but I've also found that it appears random *in* the output. For example:

```
[ERROR   ] JSON Render failed for: {
    "local": {
        "fun_args": [
            "/var/lib/acme/salt//salt_state.tgz", 
            "test=None", 
            "pkg_sum=953bb413acfa01c1e9a65c7e5258f449", 
            "hash_type=md5"
        ], 
        "jid": "20151204095708560039", 
        "return": {
   ... snip ...
            "file_|-/var/www/vhosts_|-/var/www/vhosts_|-directory": {
                "comment": "Directory /var/www/vhosts is in the correct state", THEPASSWORD
    
                "name": "/var/www/vhosts", 
                "start_time": "09:57:34.040038", 
                "result": true, 
                "duration": 0.823, 
                "__run_num__": 52, 
                "changes": {}
            }, 
    ... snip ...
        }, 
        "retcode": 0, 
        "fun": "state.pkg", 
        "id": "SID1000000", 
        "out": "highstate"
    }
}
THEPASSWORD
```

I've tried removing the thin dir on the remote, and also disabling keeping the thin dir just in case, but it still happens. I'm not sure what triggers it but I'm unable to make it stop doing this! There's nothing else that stands out in the debug output that I believe it related to this.

Versions info:

```
Salt Version:
           Salt: 2015.8.3
 
Dependency Versions:
         Jinja2: 2.8
       M2Crypto: Not Installed
           Mako: 1.0.2
         PyYAML: 3.11
          PyZMQ: Not Installed
         Python: 2.7.10 (default, Oct 14 2015, 16:09:02)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: Not Installed
           cffi: 1.1.2
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.3.4
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed
 
System Versions:
           dist: Ubuntu 15.10 wily
        machine: x86_64
        release: 4.2.0-19-generic
         system: Ubuntu 15.10 wily
```

This seems to be triggering on our use of the mysql_user state.

I added a log line at L360, so when the regex matches and it sends the password over, it dumps the `buff` variable so we can see what it's matching on:

```
BUFFDUMP: Sending password due to regex match in buff: s... yes\nchecking whether to build shared libraries... yes\nchecking whether to build static libraries... no\nconfigure: creating ./config.status\nconfig.
status: creating config.h\nconfig.status: executing libtool commands\nrunning: make\n/bin/sh /var/tmp/pear-build-rootUGb0Zr/memcache-3.0.8/libtool --mode=compile cc -I/usr/include/php -I. -I/var/tmp/memcach
e -DPHP_ATOM_INC -I/var/tmp/pear-build-rootUGb0Zr/memcache-3.0.8/include -I/var/tmp/pear-build-rootUGb0Zr/memcache-3.0.8/main -I/var/tmp/memcache -I/usr/include/php -I/usr/include/php/main -I/usr/include/ph
p/TSRM -I/usr/include/php/Zend -I/usr/include/php/ext -I/usr/include/php/ext/date/lib  -DHAVE_CONFIG_H  -g -O2   -c /var/tmp/memcache/memcache.c -o memcache.lo\nlibtool: compile:  cc -I/usr/include/php -I. 
-I/var/tmp/memcache -DPHP_ATOM_INC -I/var/tmp/pear-build-rootUGb0Zr/memcache-3.0.8/include -I/var/tmp/pear-build-rootUGb0Zr/memcache-3.0.8/main -I/var/tmp/memcache -I/usr/include/php -I/usr/include/php/main
 -I/usr/include/php/TSRM -I/usr/include/php/Zend -I/usr/include/php/ext -I/usr/include/php/ext/date/lib -DHAVE_CONFIG_H -g -O2 -c /var/tmp/memcache/memcache.c  -fPIC -DPIC -o .libs/memcache.o\nIn file inclu
ded from /var/tmp/memcache/php_memcache.h:34:0,\n                 from /var/tmp/memcache/memcache.c:30:\n/var/tmp/memcache/memcache_pool.h:45:47: fatal error: ext/standard/php_smart_str_public.h: No such fi
le or directory\n #include \"ext/standard/php_smart_str_public.h\"\n                                               ^\ncompilation terminated.\nmake: *** [memcache.lo] Error 1\nERROR: `make' failed"
                }, 
                "__id__": "install_memcache_beta_pecl"
            }, 
            "mysql_user_|-create_mysql_user_for_web_server_|-testuser_|-present": {
                "comment": "MySQL Error 1045: Access denied for user 'root'@'127.0.0.1' (using password: YES)", 
                "name": "testuser", 
                "start_time": "12:02:53.078984", 
                "result": false, 
                "duration": 0.835, 
     ENDBUFFDUMP
```

It seems to match on the JSON bit in `buff`?

The state that's being executed here is:

```
create_mysql_user_for_web_server:
  mysql_user.present:
    - name: {{ pillar['magento_domain_user']}}
    - host: {{ pillar['magento_stack_web_server_ip'] }}
    - password: {{ pillar['magento_domain_mysql_password'] | yaml_encode }}
    - connection_user: root
    - connection_host: {{ pillar['magento_domain_mysql_host'] }}
{% if pillar['magento_stack_mysql_root_password'] is defined %}
    - connection_pass: {{ pillar['magento_stack_mysql_root_password'] | yaml_encode }}
{% endif %}
```

--- a/salt/client/ssh/shell.py
+++ b/salt/client/ssh/shell.py
@@ -23,6 +23,10 @@ log = logging.getLogger(__name__)
 SSH_PASSWORD_PROMPT_RE = re.compile(r'(?:.*)[Pp]assword(?: for .*)?:', re.M)
 KEY_VALID_RE = re.compile(r'.*\(yes\/no\).*')
 
+# Keep these in sync with ./__init__.py
+RSTR = '_edbc7885e4f9aac9b83b35999b68d015148caf467b78fa39c05f669c0ff89878'
+RSTR_RE = re.compile(r'(?:^|\r?\n)' + RSTR + r'(?:\r?\n|$)')
+
 
 class NoPasswdError(Exception):
     pass
@@ -339,6 +343,7 @@ class Shell(object):
                 stream_stdout=False,
                 stream_stderr=False)
         sent_passwd = 0
+        send_password = True
         ret_stdout = ''
         ret_stderr = ''
         old_stdout = ''
@@ -353,7 +358,10 @@ class Shell(object):
                     buff = stdout
                 if stderr:
                     ret_stderr += stderr
-                if buff and SSH_PASSWORD_PROMPT_RE.search(buff):
+                if buff and RSTR_RE.search(buff):
+                    # We're getting results back, don't try to send passwords
+                    send_password = False
+                if buff and SSH_PASSWORD_PROMPT_RE.search(buff) and send_password:
                     if not self.passwd:
                         return '', 'Permission denied, no authentication information', 254
                     if sent_passwd < passwd_retries:

